node/src/main/scala/coop/rchain/node/web/WebApiRoutes.scala

```diff
@@ -0,0 +1,130 @@
+package coop.rchain.node.web
+
+import cats.effect.Sync
+import cats.syntax.all._
+import cats.~>
+import coop.rchain.node.api.WebApi
+import coop.rchain.node.api.WebApi._
+import org.http4s.{HttpRoutes, Response}
+
+object WebApiRoutes {
+  import WebApiSyntax._
+
+  def service[F[_]: Sync, M[_]: Sync](
+      webApi: WebApi[M]
+  )(implicit mf: M ~> F): HttpRoutes[F] = {
+    import io.circe.generic.auto._
+    import io.circe.syntax._
+    import org.http4s.circe._
+    import org.http4s.{EntityDecoder, EntityEncoder, InvalidMessageBodyFailure, Request}
+
+    val dsl = org.http4s.dsl.Http4sDsl[F]
+    import dsl._
+
+    def handleRequest[A, B](req: Request[F], f: A => M[B])(
+        implicit decoder: EntityDecoder[F, A],
+        encoder: EntityEncoder[F, B]
+    ): F[Response[F]] =
+      req
+        .attemptAs[A]
+        .value
+        .flatMap(_.liftTo[F])
+        .flatMap(a => mf(f(a)))
+        .flatMap(Ok(_))
+        .handleErrorWith {
+          // The place where all API errors are handled
+          // TODO: introduce error codes
+          // Request JSON parse errors
+          case err: InvalidMessageBodyFailure =>
+            BadRequest(s"${err.getMessage.take(250)}...".asJson)
+          // Errors from BlockAPI
+          case err: BlockApiException => BadRequest(err.getMessage.asJson)
+          case err: Throwable         => BadRequest(err.getMessage.asJson)
+        }
+
+    implicit class MEx[A](val ma: M[A]) {
+      // Handle GET requests
+      //   case GET -> Root / "lastFinalizedBlock" =>
+      //     webApi.lastFinalizedBlock.handle
+      def handle(implicit encoder: EntityEncoder[F, A]): F[Response[F]] =
+        mf(ma)
+          .flatMap(Ok(_))
+          .handleErrorWith(err => BadRequest(err.getMessage.asJson))
+    }
+
+    implicit class RequestEx(val req: Request[F]) {
+      // Handle POST requests
+      //   case req @ POST -> Root / "deploy" =>
+      //     req.handle[SignedDeploy, String](WebApi.toSigned[M](_) >>= webApi.deploy)
+      def handle[A, B](
+          f: A => M[B]
+      )(implicit decoder: EntityDecoder[F, A], encoder: EntityEncoder[F, B]): F[Response[F]] =
+        handleRequest[A, B](req, f)
+
+      // Handle POST requests without input parameters
+      //   case req @ POST -> Root / "lastFinalizedBlock" =>
+      //     req.handle_(webApi.lastFinalizedBlock)
+      def handle_[B](
+          f: M[B]
+      )(implicit encoder: EntityEncoder[F, B]): F[Response[F]] =
+        handleRequest[Unit, B](req, _ => f)
+    }
+
+    // TODO: Create generic encoders/decoders for
+    // ADT's with discriminator field
```

Comment:
@marcin-rzeznicki here is the example of JSON encoding.
```js
final case class ExprInt(data: Long) extends RhoExpr

// Default _circe_ encoding
{
  ExprInt: {
    data: 42
  }
}
// It would be nicer with `type` as discriminator
{
  type: "ExprInt",
  data: 42
}
```
I was trying to use `circe-generic-extras` to encode/decode case classes with discriminator field. I'm not sure how to create generic instances for _circe_ Encoder/Decoder types which are used in `http4s` EntityEncoder/EntityDecoder types.
My understandig is that _circe_ has generic instances for conversion of ADT's to HList of tuples (fields).
https://immutables.pl/2017/02/25/customizing-circes-auto-generic-derivation#custom-configuration

Another option that I tried is _semi-auto_ with manual definition of `deriveEnumerationEncoder`s.
https://dzone.com/articles/5-useful-circe-feature-you-may-have-overlooked

Also providing implicit configuration for `circe` in this context gives an error.

Reply:
```
    implicit val encodeBlockInfo: Encoder[BlockInfo] = {
      Encoder.forProduct2[BlockInfo, LightBlockInfo, Seq[DeployInfo]]("blockInfo", "deploys") {
        case BlockInfo(lb, deploys) => (lb.get, deploys)
      }
    }
```

https://circe.github.io/circe/codecs/semiauto-derivation.html

Reply:
https://github.com/circe/circe/issues/541#issuecomment-310952006

Reply:
@zsluedem thanks for improvement with encoders.